File: 2016-04_99-bottles-of-beer/erlang_elixir_roger-simms/bottles.exs
defmodule Bottles do
  def start, do: run(99)

  defp run(-1), do: :ok
  defp run(num) do
    line1(num)
    line2(num-1)
    run(num-1)
  end

  defp count_bottles(0), do: "no"
  defp count_bottles(num), do: Integer.to_string(num)

  defp pluralise(num), do: if num != 1, do: "s"

  defp line1(num), do: IO.puts "#{String.capitalize(count_bottles(num))} bottle#{pluralise(num)} of beer on the wall, #{count_bottles(num)} bottle#{pluralise(num)} of beer."
  defp line2(-1), do: IO.puts "Go to the store and buy some more, 99 bottles of beer on the wall.\n"
  defp line2(num), do: IO.puts "Take one down and pass it around, #{count_bottles(num)} bottle#{pluralise(num)} of beer on the wall.\n"
end

# Bottles.start()

File: 2016-04_99-bottles-of-beer/erlang_elixir_roger-simms/bottles.erl
-module(bottles).
-export([start/0]).


start() -> pass(99).

pass(-1) -> ok;
pass(N) ->
    line1(N),
    line2(N-1),
    pass(N-1).


bottle_count(0) -> "no";
bottle_count(N) -> integer_to_list(N).

capitalise_first(S) ->
    F = fun([H|T]) -> [string:to_upper(H) | string:to_lower(T)] end,
    string:join(lists:map(F, string:tokens(S, "")), "").

pluralise(1) -> '';
pluralise(_) -> "s".

line1(N) ->
    io:format("~s bottle~s of beer on the wall, ~s bottle~s of beer.~n", [capitalise_first(bottle_count(N)),
                                                                          pluralise(N),
                                                                          bottle_count(N),
                                                                          pluralise(N)]).

line2(-1) ->
    io:format("Go to the store and buy some more, 99 bottles of beer on the wall.~n~n");
line2(N) ->
    io:format("Take one down and pass it around, ~s bottle~s of beer on the wall.~n~n", [bottle_count(N),
                                                                                         pluralise(N)]).
